app/models/form.rb

```diff
@@ -154,48 +154,80 @@ def invalidated_pages(log, current_user = nil)
     pages.reject { |p| p.routed_to?(log, current_user) }
   end
 
-  def invalidated_questions(log)
-    invalidated_page_questions(log) + invalidated_conditional_questions(log)
-  end
-
-  def invalidated_page_questions(log, current_user = nil)
-    # we're already treating these fields as a special case and reset their values upon saving a log
-    callback_questions = %w[postcode_known la ppcodenk previous_la_known prevloc postcode_full ppostcode_full location_id address_line1 address_line2 town_or_city county]
-    questions.reject { |q| q.page.routed_to?(log, current_user) || q.derived? || callback_questions.include?(q.id) } || []
-  end
+  def reset_not_routed_questions_and_invalid_answers(log)
+    reset_not_routed_checkbox_questions(log)
 
-  def reset_not_routed_questions(log)
-    enabled_questions = enabled_page_questions(log)
-    enabled_question_ids = enabled_questions.map(&:id)
+    reset_radio_questions_if_not_routed_or_invalid_answers(log)
 
-    invalidated_page_questions(log).each do |question|
-      if %w[radio checkbox].include?(question.type)
-        enabled_answer_options = enabled_question_ids.include?(question.id) ? enabled_questions.find { |q| q.id == question.id }.answer_options : {}
-        current_answer_option_valid = enabled_answer_options.present? ? enabled_answer_options.key?(log.public_send(question.id).to_s) : false
+    reset_not_routed_free_user_input_questions(log)
+  end
```

Comment:
the above method is a development on `reset_not_routed_questions` and now delegates to methods for questions of each type